python/parameters/Parameter.py

```diff
@@ -138,108 +165,126 @@ def private(self):
         return self.__private
 
     @default.setter
-    def default(self, value):
+    def default(self, val):
         """Set the default value for this option."""
-        self.__default = value
-        if self.__value is None:
-            self.value = self.__default
+        if val is None:
+            self.__default = None
+
+        else:
+            d = self.__check(val)
+            if d is not None:
+                self.__default = d
+                if self.__value is None:
+                    self.__value = self.__default
 
     @value.setter
     def value(self, val):
         """
         Sets the value and performs a myriad of consistency checks.
         """
-
-        if val is None:
+        if (val is None) and (self.__value is not None):
             self.__value = None
+            self.__set_by_user = True
             return
 
+        v = self.__check(val)
+        if v is not None:
+            self.__value = v
+            self.__set_by_user = True
+
+    def isSetByUser(self):
+        """Return True if the value has been set after construction."""
+        return self.__set_by_user
+
+    def validate(self):
+        """Validate that the Parameter is in the correct state."""
+        if self.__required and (self.value is None):
+            msg = "The Parameter '%s' is marked as required, but no value is assigned."
+            LOG.warning(msg, self.name)
+            return 1
+        return 0
+
+    def toString(self, prefix='', level=0):
+        """Create a string of Parameter information."""
+        from .InputParameters import InputParameters
+        is_sub_option = (self.__vtype is not None) and (InputParameters in self.__vtype) and (self.__value is not None)
+
+        out = [mooseutils.colorText(self.__name, 'LIGHT_YELLOW')]
+        if prefix is not None:
+            out[0] = '{} | {}{}'.format(out[0], prefix, self.__name) if prefix else out[0]
+
+        if self.__doc is not None:
+            wrapper = textwrap.TextWrapper()
+            wrapper.initial_indent = ' '*2
+            wrapper.subsequent_indent = ' '*2
```

Comment:
Made me laugh =D

I suppose this is the standard way to create two spaces? I learned something today.